File: 3211. Generate Binary Strings Without Adjacent Zeros.cpp
class Solution {
public:
    void backtrack(vector<string>& ans, string& curr, bool lastZero, int n) {
        if (curr.length() == n) {
            ans.push_back(curr);
            return;
        }

        if (!lastZero) {
            curr.push_back('0');
            backtrack(ans, curr, true, n);
            curr.pop_back();
        }

        curr.push_back('1');
        backtrack(ans, curr, false, n);
        curr.pop_back();
    }

    vector<string> validStrings(int n) {
        vector<string> ans;
        string curr;
        backtrack(ans, curr, false, n);
        return ans;
    }
};

File: 3211. Generate Binary Strings Without Adjacent Zeros.java
class Solution {
    public void backtrack(List<String> ans, StringBuilder curr, boolean lastZero, int n){
        if(curr.length() == n){
            ans.add(curr.toString());
            return;
        }

        if(!lastZero){
            curr.append('0');
            backtrack(ans, curr, true, n);
            curr.deleteCharAt(curr.length() - 1);
        }
        curr.append('1');
        backtrack(ans, curr, false, n);
        curr.deleteCharAt(curr.length() - 1);
    }
    public List<String> validStrings(int n) {
        List<String> ans = new ArrayList<>();
        backtrack(ans, new StringBuilder(),false, n);

        return ans;
    }
}